pkg/resolution/resolver/framework/controller_v2.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"strings"
+
+	rrclient "github.com/tektoncd/pipeline/pkg/client/resolution/injection/client"
+	rrinformer "github.com/tektoncd/pipeline/pkg/client/resolution/injection/informers/resolution/v1beta1/resolutionrequest"
+	"github.com/tektoncd/pipeline/pkg/resolution/common"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/clock"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+// ReconcilerModifierV2 is a func that can access and modify a reconciler
+// in the moments before a resolver is started. It allows for
+// things like injecting a test clock.
+type ReconcilerModifierV2 = func(reconciler *ReconcilerV2)
+
+// NewControllerV2 returns a knative controller for a Tekton Resolver.
+// This sets up a lot of the boilerplate that individual resolvers
+// shouldn't need to be concerned with since it's common to all of them.
+func NewControllerV2(ctx context.Context, resolver ResolverV2, modifiers ...ReconcilerModifierV2) func(context.Context, configmap.Watcher) *controller.Impl {
```

Comment:
Worth calling out that the compatibility concern for the original Resolver interface ends here, since the actual implementation of the reconciler is unexported.

You could use this as a means for backwards compatibilty into v1 here by simplifying the problem as instead of going from v1 to v2, going from 2 different resolver types in the same version. This might solve some of the "v2 having a v1 API" issues we're seeing elsewhere.

e.g. something like:

```go
func NewController(ctx context.Context, resolver Resolver, modifiers ...ReconcilerModifier) func(context.Context, configmap.Watcher) *controller.Impl {
...
}

func NewRequestController(ctx context.Context, resolver RequestResolver, modifiers ...ReconcilerModifier) func(context.Context, configmap.Watcher) *controller.Impl {
...
}
```

I think this should be doable, since you should always be able to wrap this in a backwards compatible way:

```go
type compatResolver interface {
  old Resolver
}

func (*compatResolver) ValidateRequest(ctx context.Context, req *v1beta1.ResolutionRequestSpec) error {
  // if any field but params set, error
}

func (r *compatResolver) ResolveRequest(ctx context.Context, req *v1beta1.ResolutionRequestSpec) (ResolvedResource, error) {
  return r.Resolve(req.Params)
}